Хорошая работа, все необходимые файлы есть хорошая структура проекта. В целом хорошо, нужно исправить комментарии(Надо исправить) на данной итерации.

**Надо исправить**
- Код вернул `"password": "$2a$10$dsrTmlbHD7Z0jO4t.CYmCe829tf1C59vvIzhOz.J/6BGxTtP06bTa"` при регистрации. API не должен возвращать хеш пароля. Хеш должен быть приватным.
- В случае повторной регистрации email на тот же адрес необходимо возвращать ошибку со статус-кодом 409 (Conflict) вместо 500й. Можно отловить ошибку при помощи значений err.name = MongoError и err.code = 11000.
- Исправить ошибки линтинга (их можно увидеть запустив команду npx eslint ./)
```
/Mesto-nodeJS-mongo-withAuth/middlewares/auth.js
  1:1  error  Parsing error: The keyword 'const' is reserved

✖ 1 problem (1 error, 0 warnings)
```
- Нужно исправить возможно зарегистрировать пользователя с пустым паролем. Обратите внимание, что вы сохраняете в БД уже хэш пароля (который даже от пустой строки будет необходимой длины), поэтому проходите ограничение на минимальную длину.
- Нужно исправить возможно зарегистрировать пользователя с паролем состоящий только из пробелов.



**Можно лучше**
- Секретный ключ для выпуска jwt токена и для верификации жестко прописан в разных местах кода (key), а не реквайрится из одного места; Лучше брать секретный ключ из env переменных или конфиг файла. Для env переменных `const { JWT_SECRET } = process.env;` лучше указать значение по умолчания в случае если не будет env переменных. Пример `const { JWT_SECRET = 'dev-key' } = process.env;`
- Добавить аттрибут sameSite для увеличения безопасности передачи cookies. 
```
 .cookie('jwt', token, {
      httpOnly: true,
      sameSite: true
})
```

--- a/README.md
+++ b/README.md
@@ -26,6 +26,7 @@ NodeJS, express, REST API
   - "jsonwebtoken": "^8.5.1"
   - "validator": "^13.1.17"
   - "bcrypt": "^5.0.0"
+  - "mongoose-unique-validator": "^2.0.3"
 
 ## Инструкции по запуску:
 - Скачать или склонировать репозиторий

--- a/controllers/users.js
+++ b/controllers/users.js
@@ -79,7 +79,7 @@ module.exports.createUser = (req, res) => {
     .catch((err) => {
       if (err.errors.email.message === 'Error, expected value to be unique.') {
         res.status(409).send({ message: "Переданы некорректные данные" });
-      }
+      } else
       if (err.name === "ValidationError") {
         res.status(400).send({ message: "Переданы некорректные данные" });
       } else {

--- a/controllers/users.js
+++ b/controllers/users.js
@@ -62,13 +62,22 @@ module.exports.updateAvatar = (req, res) => {
 };
 
 module.exports.createUser = (req, res) => {
+  const {email, password, name, about, avatar} = req.body;
+  if (!password || password.length < 8 || /\s/.test(password)) {
+    const password = null;
+    return res.status(400).send({ message: 'некорректный пароль' });
+  }
   bcrypt.hash(req.body.password, 10)
     .then((hash) => {
-      const { name, about, avatar, email } = req.body;
       return UserModel.create({ name, about, avatar, email, password: hash })
     })
-    .then(user => res.send({ data: user }))
+    .then((user) => {
+      const { name, about, avatar, email, _id } = user
+      res.send({ data: { name, about, avatar, email, _id } })})
     .catch((err) => {
+      if (err.errors.email.message === 'Error, expected value to be unique.') {
+        res.status(409).send({ message: "Переданы некорректные данные" });
+      }
       if (err.name === "ValidationError") {
         res.status(400).send({ message: "Переданы некорректные данные" });
       } else {
@@ -85,7 +94,7 @@ module.exports.login = (req, res) => {
       // создадим токен
       const token = jwt.sign({ _id: user._id }, 'some-secret-key', { expiresIn: '7d' });
       // вернём токен
-      res.cookie('jwt', token, { maxAge: 3600000 * 24 * 7, httpOnly: true }).end();
+      res.cookie('jwt', token, { maxAge: 3600000 * 24 * 7, httpOnly: true, sameSite: true}).end();
     })
     .catch((err) => {
       if (err.message === "Неправильные почта или пароль") {

--- a/controllers/users.js
+++ b/controllers/users.js
@@ -62,13 +62,15 @@ module.exports.updateAvatar = (req, res) => {
 };
 
 module.exports.createUser = (req, res) => {
-  const {email, password, name, about, avatar} = req.body;
+  let { password } = req.body;
   if (!password || password.length < 8 || /\s/.test(password)) {
-    const password = null;
+    
     return res.status(400).send({ message: 'некорректный пароль' });
   }
+  password = null;
   bcrypt.hash(req.body.password, 10)
     .then((hash) => {
+      const {email, name, about, avatar} = req.body;
       return UserModel.create({ name, about, avatar, email, password: hash })
     })
     .then((user) => {

--- a/middlewares/auth.js
+++ b/middlewares/auth.js
@@ -1,3 +1,4 @@
+/*eslint-env es6*/
 const jwt = require('jsonwebtoken');
 
 module.exports = (req, res, next) => {

--- a/models/user.js
+++ b/models/user.js
@@ -1,6 +1,7 @@
 /*eslint-env es6*/
 const mongoose = require('mongoose');
 const isEmail = require('validator/lib/isEmail');
+const uniqueValidator = require('mongoose-unique-validator')
 const bcrypt = require('bcrypt');
 
 const userSchema = new mongoose.Schema({
@@ -42,6 +43,8 @@ const userSchema = new mongoose.Schema({
     select: false
   },
 });
+
+userSchema.plugin(uniqueValidator, { message: 'Error, expected value to be unique.' });
 
 userSchema.statics.findUserByCredentials = function (email, password) {
   return this.findOne({ email }).select('+password')

--- a/package.json
+++ b/package.json
@@ -22,6 +22,7 @@
     "express-rate-limit": "^5.1.3",
     "jsonwebtoken": "^8.5.1",
     "mongoose": "^5.10.7",
+    "mongoose-unique-validator": "^2.0.3",
     "validator": "^13.1.17"
   }
 }

